### ISSUE TITLE
S7 figure

### ISSUE THREAD

**Issue by scharch** [psathyrella/selection-metric-comments#3]
In the right panels, I don't understand how the hamming distance for the nt consensus can be so high while being essentially zero for the aa consensus. In the left panels, the difference looks to be roughly a factor of 2.5, which is much more in line with what I would expect.

**Comment by psathyrella** [comment]
I agree this is weird, and I didn't think about it a ton when I made the plot, but I've been assuming it was from something like fixed synonymous mutations (maybe hitchiking is an appropriate word to use here?). Very open to other thoughts though. I had classed it as 'weird but not impossible', but if you feel it's more in the 'impossible' realm I can spend some more time poking at it.

**Comment by scharch** [comment]
I lean more toward the impossible, but I do see one possibility: if this observation time is so long/late that essentially everything is at the target sequence(s), then that could result in a scenario like you describe. Of course you might want to exclude that from the testing regime for other reasons...

**Comment by psathyrella** [comment]
Yes, I think it's likely that this is what's happening, although I don't have the simulation truth diversity plots at my fingertips to really check. I think the intuition is that at all not-super-late times, the family is "stretched out" in the direction of the target sequence, between where their consensus used to be, and where it's going. Whereas given infinite time the family's consensus of course reaches the target sequence, at which point there's nowhere to go, so instead of being bimodal or something, the family is just fuzz (in nucleotide space) around the consensus sequence, which (in AA space) is now also the target sequence. Maybe a better way to say it is that this is showing us why aa-cdist is so much better than nuc-cdist and lbi -- the inaccuracy in aa-cdist at earlier times is only because the family is still evolving toward the target sequence (and not because of random noise, which is effectively removed by the AA consensus procedure). So at later times when the inaccuracy in aa-cdist due to the family being stretched out is removed, we can finally see how little error there was in aa-cdist just due to random chance.

I think this is why you see at very large times the aa-cdist inaccuracy going to implausibly low values (S2 Figure):
![p](https://user-images.githubusercontent.com/2960243/81018711-e7a68d00-8e19-11ea-9b4d-ea568cff5954.png)

This is what I was getting at here:
![p](https://user-images.githubusercontent.com/2960243/81021137-563a1980-8e1f-11ea-8f4f-dc81c5290c27.png)


While I partially agree with excluding times that are large enough that the family has converged to the target sequence, since they might represent parts of parameter space that are not well described by our simulation model, I also think it might be more clear to include them while being more clear what's going on. Because I think it's likely that there's not anything fundamentally wrong with the dynamics of the model at late times, rather that there's inevitably just fewer sources of noise than in real data (e.g. we don't model position-biased sequencing error), so instead of plateauing at some small inaccuracy value, the simulation says it plateaus at zero. Stopping at earlier times will make the plots look better (since they won't have implausibly optimistic numbers), but nothing really changes about the dynamics, e.g. in S7 Figure we'd still conclude that aa-cdist accuracy improves at later times, we just wouldn't conclude that it actually goes to zero (but we already know it would never really go to zero).

In the end, as I allude to in the screenshot ^, I'm not sure what should be done with "optimal antibodies"/target sequences. It would be easy to just never use simulation where the families get close to the target sequences. But I don't know of any reason why there shouldn't in some cases exist optimal antibodies in the sense of strong local maxima in fitness (see ref in that screenshot, as well).

**Comment by scharch** [comment]
Rereading that paragraph in the light of this conversation, I think I see part of why I missed it on the first read: 
> the family consensus is an extremely accurate predictor of the hypothetical optimal antibody toward which selection is pushing

...it's not _predicting_ the optimal sequence, it _is_ the optimal sequence! And I think that you need to unpack that reference a bit more, especially since it's a preprint that a lot of people probably won't have seen yet. (Thanks for flagging it, BTW, it's an interesting piece.) 

Moreover, I'm not convinced that it's just a question of whether/how often a "ceiling" scenario arises _in vivo_ --I actually think there's an important difference between the simulation and biology in this scenario. The sim fixes one (or a few) amino acid sequences; once these are reached, further SHM is subjected to stringent purifying selection. By contrast, the definition of the ceiling scenario is that there are _many_ equally fit sequences (even within a single local maximum), which should result in substantial drift and diversity. In fact, I wouldn't be surprised if the result ends up being a lack of a strong consensus. I suppose it wouldn't be terribly difficult to simulate this (just treat everything with Hamming distance <= 5 as equally fit), though I'm not suggesting that I think it's in any way necessary. I guess that's all just what I mean by unpacking ref20...

**Comment by psathyrella** [comment]
oh damn that's a really good point. What I was meaning above was the simulation clearly didn't have enough fuzz, but I couldn't think of a sensible way to make it fuzzier.

But yeah, it totally seems more sensible to have affinity depend on similarity to only some, rather than all, residues in the target sequence. E.g. decide cdr3 contacts the antigen, so let cdr1/2 be fairly unconstrained, while fwk can't change much at all. That would be easy to implement (although I don't know that I'll do it now, already submitted the current version), but definitely worth mentioning in the text in future versions.

**Comment by psathyrella** [comment]
I actually misinterpreted your suggestion slightly, so ended up implementing this in two different ways. But doing it the simple way that I think you meant (just stop increasing affinity once a sequence gets closer to target than some threshold), here's how the right side of S7 Fig (large carry cap and obs time) looks for thresholds of 0 (left, same as first draft), 2 (center, used for second draft), and 5 (right) amino acid changes:

![cdist-accuracy-vs-min-target-distance](https://user-images.githubusercontent.com/2960243/86389442-29518980-bc4b-11ea-857f-0341d7309aaa.png)

**Comment by scharch** [comment]
Interesting, thanks. I'm a little surprised it's still so constrained... Why is the nucleotide consensus more accurate at low N with the affinity cap in place?

**Comment by psathyrella** [comment]
Yeah i can't think of a reason that would be the case; I think it's more likely that I'm underestimating my uncertainties somehow. It's a little bit complicated to get each one of those points, and I think I'm just using the standard error over different families, so if something just isn't quite independent that's supposed to be, that'd do it. In some of those plots there's clearly either some fine structure, or the uncertainties are a bit too small (one reason there could be fine structure is the even vs odd N sampled affecting tie resolution in the consensus, i.e. maybe odd N sampled are more accurate because they have fewer ties, or the oppose).

Also, if we just remove that one point near 15 in the top left plot, they look much more compatible. Also also, whatever happens from 0 --> 2 I would expect to happen from 2 --> 5, but the top right two plots are pretty similar.

And I spent a lot of time staring at the individual family plots [like this](https://user-images.githubusercontent.com/2960243/82702661-45631380-9c27-11ea-88ac-22a0f7e594de.png) when i implemented the threshold, and the nucleotide evolution didn't seem very affected by the threshold, it just kept on trucking when the family ran up against the AA threshold (which I think makes sense -- I think it's another example of the extreme high dimensionality of this space foiling my three-dimensional intuition, i.e. there's 400 different directions for evolution to go in).
